pkg/api/dialer.go

```diff
@@ -0,0 +1,100 @@
+package api
+
+import (
+	"context"
+	"errors"
+	"net"
+	"os"
+	"strings"
+	"testing"
+	"time"
+)
+
+// DialContextFunc is a function that dials a context, network, and address.
+type DialContextFunc func(ctx context.Context, network, address string) (net.Conn, error)
+
+// RetryDialUntilSuccess will retry every `retryTimeout` until it succeeds.
+func RetryDialUntilSuccess(retryTimeout time.Duration) DialContextFunc {
+	return func(ctx context.Context, network, address string) (net.Conn, error) {
+		for {
+			dialer := &net.Dialer{
+				Timeout:   retryTimeout,
+				KeepAlive: 30 * time.Second, // Similar to the default HTTP dialer.
+			}
+			c, err := dialer.DialContext(ctx, network, address)
+			if err != nil {
+				if errors.Is(err, context.DeadlineExceeded) {
+					continue
+				}
+				if errors.Is(err, os.ErrDeadlineExceeded) {
+					continue
+				}
```

Comment:
understanding check: these are the only errors we're retrying for and others will just get returned below?